### DISCUSSION TITLE
How can I avoid putting the $browsers inside each other?

### DISCUSSION THREAD

**Discussion by AzraelD03** [reactphp/reactphp#563] [Q&A]
```
<?php

require_once __DIR__ . '/vendor/autoload.php';

use React\EventLoop\Loop;
use React\Socket\SocketServer;
use React\Http\HttpServer;
use React\Http\Browser;
use Psr\Http\Message\ServerRequestInterface;
use React\Http\Message\Response;
use Psr\Http\Message\ResponseInterface;

$socket = new SocketServer('0.0.0.0:8080');

$browser = new Browser();

$http = new HttpServer(function (ServerRequestInterface $req) use ($browser) {
    $input = json_decode((string)$req->getBody());

    $id = $input->message->chat->id;
    $msg = $input->message->text;

    $url = 'https://api.telegram.org/bot';
    $url .= 'xxx';
    $url .= '/sendMessage?';
    $url .= http_build_query([
        'chat_id' => $id,
        'text' => $msg,
        'parse_mode' => 'HTML'
    ]);


    $browser->get($url)
        ->then(function (ResponseInterface $res)  use ($browser) {
            Loop::addTimer(5, function () use ($browser, $res) {
                $output = json_decode((string)$res->getBody());

                $id = $output->result->chat->id;
                $msg_id = $output->result->message_id;

                $url = 'https://api.telegram.org/bot';
                $url .= 'xxx';
                $url .= '/editMessageText?';
                $url .= http_build_query([
                    'chat_id' => $id,
                    'message_id' => $msg_id,
                    'text' => 'Very good!',
                    'parse_mode' => 'HTML'
                ]);

                $browser->get($url)
                    ->then(function (ResponseInterface $res) {
                    }, function (Exception $e) {
                        echo 'Error: ' . $e->getMessage() . PHP_EOL;
                    });
            });
        }, function (Exception $e) {
            echo 'Error: ' . $e->getMessage() . PHP_EOL;
        });

    return Response::plaintext(
        'True'
    );
});

$http->listen($socket);
```

I would like to keep the code asynchronous, but not all $browser->methods inside each other, is it possible?

**Comment by SimonFrings** [comment]
Hey @AzraelD03, interesting question.

Answering this question depends on how you want your project to work. Taken from your code example above, you're currently sending out a HTTP `GET` request and once you receive your response, you overwrite `$url` and send out another `GET` request. This seems to me like you depend on the response from your first `GET` in order to send the second request, so this seems right to me. The same goes for your second suggestion in #564.

If you simply don't want to have this nested as it currently is, you could also use `async()` and `await()` instead of promises here by using the [ReactPHP Async component](https://github.com/reactphp/async). I gave a quick explanation for the usage of `async()` together with `await()` in [#546](https://github.com/orgs/reactphp/discussions/546#discussioncomment-7640814), so maybe this will help understanding how to use these functions properly. The rest should be covered in our [Async documentation](https://github.com/reactphp/async?tab=readme-ov-file#async).

You can also check out [Framework X](https://framework-x.org/) which uses ReactPHP underneath and provides you with a simple framework to write async PHP code. This way you don't have to deal with all the low-level stuff, but this is completely up to you. 

I hope this helps :+1:

FYI: I saw that you also opened #564 and it seems like these discussions are pretty similar, so let's use this ticket here as our main place to discuss this and I'll set the needed references in #564.
